### PR TITLE
Implement endpoint `getMerklePath`

### PR DESCRIPTION
**Pull request by kchulj**
We need the Merkle path for the Merkle proof.

Solution for #298

### CODE REVIEW

**Comment by kchulj** [comment]
I will add test code in `Stoa.test.ts`.

**Comment by kchulj** [comment]
@MichaelKim20 The method in `tests/Utils` uses data from `TestAgora` `blocks`.

File: src/modules/agora/AgoraClient.ts [hunk 1]
```diff
@@ -105,6 +105,30 @@ export class AgoraClient implements FullNodeAPI
...
+        {
+            let uri = URI("/merkle_path")
+                .addSearch("block_height", block_height.toString())
+                .addSearch("hash", hash);
+
+            this.client.get(uri.toString())
+                .then((response: AxiosResponse) =>
+                {
+                    if (response.status == 200)
+                        resolve(response.data.map((entry: any) => Block.reviver("", entry)));
```
Comment (MichaelKim20) [inline src/modules/agora/AgoraClient.ts]:
```suggestion
                        resolve(response.data.map((entry: any) => new Hash(entry)));
```

File: tests/Utils.ts [hunk 2]
```diff
@@ -206,6 +225,16 @@ export class TestAgora implements FullNodeAPI
         );
     }
 
+    /// Implements FullNodeAPI.getMerklePath
+    public getMerklePath (block_height: Height, hash: Hash): Promise<Hash[]>
+    {
+        let sample_merkle_tree = this.blocks[1].merkle_tree;
+
+        return Promise.resolve(
+            [sample_merkle_tree[1], sample_merkle_tree[9], sample_merkle_tree[13]]
```
Comment (MichaelKim20) [inline tests/Utils.ts]:
```suggestion
            [
                new Hash(sample_merkle_tree[1]),
                new Hash(sample_merkle_tree[9]),
                new Hash(sample_merkle_tree[13])
            ]
```
The `sample_merkle_tree` is the Array of string

File: src/modules/agora/AgoraClient.ts [hunk 3]
```diff
@@ -105,6 +105,30 @@ export class AgoraClient implements FullNodeAPI
...
+     * @param block_height - The height of the block
+     * @param hash - The hash of the transaction
+     */
+    public getMerklePath (block_height: Height, hash: Hash): Promise<Hash[]>
+    {
+        return new Promise<Hash[]>((resolve, reject) =>
+        {
+            let uri = URI("/merkle_path")
+                .addSearch("block_height", block_height.toString())
+                .addSearch("hash", hash);
```
Comment (MichaelKim20) [inline src/modules/agora/AgoraClient.ts]:
```suggestion
                .addSearch("hash", hash.toString());
```

File: src/modules/agora/AgoraClient.ts [hunk 4]
```diff
@@ -105,6 +105,30 @@ export class AgoraClient implements FullNodeAPI
         });
     }
 
+    /**
+     * Requests and receives the Merkle path.
+     * @param block_height - The height of the block
+     * @param hash - The hash of the transaction
+     */
+    public getMerklePath (block_height: Height, hash: Hash): Promise<Hash[]>
```
Comment (MichaelKim20) [inline src/modules/agora/AgoraClient.ts:113]: The code is not currently being tested.
Please add the test code below.

```

    it ('Test of the path /merkle_path by AgoraClient', async () =>
    {
        const agora_addr: URL = new URL('http://localhost:2826');
        let agora_client = new AgoraClient(agora_addr);
        let merkle_path: Array<Hash> = await agora_client.getMerklePath(new Height("1"), new Hash("0x535b358337d919474f3043db1f292a1ac44a4f4dbbaa6d89226c7abd96c38bf96018e67828ec539623e475d480af99499368780ae346dfb6cb048b377cbc92d0"));

        let expected =
            [
                new Hash("0x7440a29292bb43361c049168c0efacf0dc3df7aa6657af59e61642f38e2a61988e4db9f59d23bc2c47b6b92dc02c3c974e8778552274f3a55498d9fce2aa3536"),
                new Hash("0x73986a582a08631cfaece0035a0012d8574aca73813da00700a90558eca2147f78d6a1c3b6aa50ea1c4abb3c270bf34fae717eaffc5ee092a1ee4ec3803f5bfd"),
                new Hash("0x943cb81d3af818b91ab200a1a8f5820a98ba02c5b98c6f8d03589175cc257fc8008ada7aa47ee1d879a79bc04d4eab7448de42c2c5174ba066981fe03f295bf5"),
            ];

        assert.deepStrictEqual(merkle_path, expected);
    });
```

**Review by MichaelKim20 — CHANGES_REQUESTED** [review]
That's a good approach. I'd like to request some modifications.

File: tests/Stoa.test.ts [hunk 5]
```diff
@@ -894,3 +894,63 @@ describe ('Test of the path /utxo for freezing', () =>
...
+            .setSearch("block_height", "1")
+            .setSearch("hash", "0x535b358337d919474f3043db1f292a1ac44a4f4dbbaa6d89226c7abd96c38bf96018e67828ec539623e475d480af99499368780ae346dfb6cb048b377cbc92d0");
+
+        let response = await client.get(uri.toString());
+
+        let expected =
+            [
+                "0x7440a29292bb43361c049168c0efacf0dc3df7aa6657af59e61642f38e2a61988e4db9f59d23bc2c47b6b92dc02c3c974e8778552274f3a55498d9fce2aa3536",
+                "0x73986a582a08631cfaece0035a0012d8574aca73813da00700a90558eca2147f78d6a1c3b6aa50ea1c4abb3c270bf34fae717eaffc5ee092a1ee4ec3803f5bfd",
+                "0x943cb81d3af818b91ab200a1a8f5820a98ba02c5b98c6f8d03589175cc257fc8008ada7aa47ee1d879a79bc04d4eab7448de42c2c5174ba066981fe03f295bf5",
```
Comment (kchulj) [inline tests/Stoa.test.ts:952]: Should I also make these `new Hash()`?

File: tests/Utils.ts [hunk 6]
```diff
@@ -157,6 +157,25 @@ export class TestAgora implements FullNodeAPI
             }, next);
         });
 
+        this.agora.get("/merkle_path", (req: express.Request, res: express.Response, next) =>
+        {
+            if (req.query.block_height === undefined || Number.isNaN(req.query.block_height))
+                return Promise.reject(new Error('Missing or invalid query parameter block_height'));
```
Comment (MichaelKim20) [inline tests/Utils.ts]:
```suggestion
            if (req.query.block_height === undefined || Number.isNaN(req.query.block_height))
            {
                res.status(400).json({ statusText: 'Missing or invalid block_height query parameter' });
                return;
            }
```

File: tests/Utils.ts [hunk 7]
```diff
@@ -157,6 +157,25 @@ export class TestAgora implements FullNodeAPI
...
+
+            if (req.query.hash === undefined)
+                return Promise.reject(new Error('Missing query parameter enroll_key'));
```
Comment (MichaelKim20) [inline tests/Utils.ts]:
```suggestion
            if (req.query.hash === undefined)
            {
                res.status(400).json({ statusText: 'Missing or invalid hash query parameter' });
                return;
            }
```

**Review by MichaelKim20 — APPROVED** [review]
It would be nice to modify the exception processing for part 2 of the test code.

File: tests/Utils.ts (hunk 7 above)
Comment (kchulj) [reply tests/Utils.ts]: Thank you for the suggestion. Should this comment be removed?
https://github.com/bosagora/stoa/blob/534ef096aa1742749cc39d231b1fcfd20f4da494/tests/Utils.ts#L107

**Comment by kchulj** [comment]
Thank you for the review @MichaelKim20 .
I have applied the suggestions.

**Pull request merged**